src/lib/operators/index_scan.cpp

```diff
@@ -0,0 +1,154 @@
+    job_task->schedule();
+  }
+
+  CurrentScheduler::wait_for_tasks(jobs);
+
+  return _out_table;
+}
+
+void IndexScan::_on_cleanup() {}
+
+void IndexScan::_validate_input() {
+  Assert(_scan_type != ScanType::OpLike, "Scan type not supported by index scan.");
+  Assert(_scan_type != ScanType::OpNotLike, "Scan type not supported by index scan.");
+
+  Assert(_left_column_ids.size() == _right_values.size(),
+         "Count mismatch: left column IDs and right values don’t have same size.");
+  if (_scan_type == ScanType::OpBetween) {
+    Assert(_left_column_ids.size() == _right_values2.size(),
+           "Count mismatch: left column IDs and right values don’t have same size.");
+  }
+
+  Assert(_in_table->get_type() == TableType::Data, "IndexScan does only support persistent tables right now.");
+}
+
+PosList IndexScan::_scan_chunk(const ChunkID chunk_id) {
+  const auto to_row_id = [chunk_id](ChunkOffset chunk_offset) { return RowID{chunk_id, chunk_offset}; };
+
+  auto range_begin = BaseIndex::Iterator{};
+  auto range_end = BaseIndex::Iterator{};
+
+  const auto& chunk = _in_table->get_chunk(chunk_id);
+  auto matches_out = PosList{};
+
+  const auto index = chunk.get_index_for(_index_type, _left_column_ids);
+  Assert(index != nullptr, "Index of specified type not found for column (vector).");
+
+  switch (_scan_type) {
+    case ScanType::OpEquals: {
+      range_begin = index->lower_bound(_right_values);
+      range_end = index->upper_bound(_right_values);
+      break;
+    }
+    case ScanType::OpNotEquals: {
+      // first, get all values less than the search value
+      range_begin = index->cbegin();
+      range_end = index->lower_bound(_right_values);
+
+      matches_out.reserve(std::distance(range_begin, range_end));
+      std::transform(range_begin, range_end, std::back_inserter(matches_out), to_row_id);
+
+      // set range for second half to all values greater than the search value
+      range_begin = index->upper_bound(_right_values);
+      range_end = index->cend();
+      break;
+    }
+    case ScanType::OpLessThan: {
+      range_begin = index->cbegin();
+      range_end = index->lower_bound(_right_values);
+      break;
+    }
+    case ScanType::OpGreaterThanEquals: {
+      range_begin = index->lower_bound(_right_values);
+      range_end = index->cend();
+      break;
+    }
+    case ScanType::OpLessThanEquals: {
+      range_begin = index->cbegin();
+      range_end = index->upper_bound(_right_values);
+      break;
+    }
+    case ScanType::OpGreaterThan: {
+      range_begin = index->upper_bound(_right_values);
+      range_end = index->cend();
+      break;
+    }
+    case ScanType::OpBetween: {
+      range_begin = index->lower_bound(_right_values);
+      range_end = index->upper_bound(_right_values2);
+      break;
+    }
+
+    default:
+      Fail("Unsupported comparison type encountered");
+  }
+
+  matches_out.reserve(matches_out.size() + std::distance(range_begin, range_end));
```

Comment:
Is `std::distance` supported in O(1) in the indexes? Otherwise, I am not sure if this `reserve` actually improves the performacne.

Reply:
Yes, if the iterators are random access iterators, which they are in this case.